File: src/Solutions/Euler244.hs
{-# LANGUAGE Strict #-}
module Solutions.Euler244 where
import Data.Array
import Data.Char(ord, intToDigit)
import Data.List(delete, findIndex, findIndices)
import Data.Maybe(catMaybes, fromJust)
import qualified Data.IntMap.Strict as IntMap
import qualified Data.IntSet as IntSet
import Numeric(showIntAtBase)
import Utils.List(ascendingMinus, ascendingSum, foldDescend)
{-
-------------
| 0| 1| 2| 3|
-------------
| 4| 5| 6| 7|
-------------
| 8| 9|10|11|
-------------
|12|13|14|15|
-------------

nodes ~ positions
vertices ~ moves
Number of vertices in graph that we are investigating is 16 * binom(15, 7) = 102960
Each node has at most 4 edges.
-}

data Position = Position
    { blank :: Int
    , reds :: [Int]
    }

-- Encoding position as int: \sum_{i = 0}^15 a_i3^i, where a_i depends on colour of the field:
--  * blue -> 0
--  * red -> 1
--  * blank -> 2

blues :: Position -> [Int]
blues pos = ascendingMinus (delete (blank pos) [0..15]) (reds pos)

isBlue, isRed, isBlank :: Int -> Position -> Bool
isBlue n pos = n `elem` blues pos
isRed n pos = n `elem` reds pos
isBlank n pos = n == blank pos

posToIndex :: Position -> Int
posToIndex (Position b rs) = 2 * 3 ^ b + sum (map (3^) rs)

indexToPos :: Int -> Position
indexToPos n = Position (fromJust $ findIndex (=='2') s) (findIndices (=='1') s) where
    s' = showIntAtBase 3 intToDigit n ""
    s = reverse $ replicate (16 - length s') '0' ++ s'

move exclCond newBlank pos@(Position b rs)
    | exclCond b = Nothing
    | otherwise = Just $ Position b' $ if isBlue b' pos then rs else ascendingSum [b] (delete b' rs)
    where b' = newBlank b

-- Move red/blue tile in the direction (moves blank in the opposite direction)
moveUp, moveDown, moveLeft, moveRight :: Position -> Maybe Position
moveUp = move (>11) (+4)
moveDown = move (<4) (subtract 4)
moveLeft = move (\b -> b `mod` 4 == 3) succ
moveRight = move (\b -> b `mod` 4 == 0) pred

neighboursPos :: Position -> [Position]
neighboursPos pos = catMaybes $ map ($ pos) [moveUp, moveDown, moveLeft, moveRight]

neighboursIndexes :: Int -> [Int]
neighboursIndexes = map posToIndex . neighboursPos . indexToPos 

initPos, finalPos :: Position
initPos = Position 0 [1, 4, 5, 8, 9, 12, 13]
finalPos = Position 0 [2, 5, 7, 8, 10, 13, 15]
initIndex, finalIndex :: Int
initIndex = posToIndex initPos 
finalIndex = posToIndex finalPos

-- Assuming that second position was created from the first position by making one move,
-- find that move and return ASCII value of the move.

findMove :: Position -> Position -> Int
findMove prePos postPos = case blank postPos - blank prePos of
    4 -> ord 'U'
    -4 -> ord 'D'
    1 -> ord 'L'
    -1 -> ord 'R'

updateChecksum :: Int -> Int -> Int
updateChecksum checksum ord = (243 * checksum + ord) `mod` 100000007

positionsBFS :: IntMap.IntMap Int
positionsBFS = go (IntSet.singleton initIndex) 1 (IntMap.singleton initIndex 0) where
    go poss i degs
        | IntSet.member finalIndex poss = IntMap.insert finalIndex i degs
        | otherwise = go nposs (i + 1) ndegs
        where nposs = IntSet.difference (IntSet.fromList $ concatMap neighboursIndexes $ IntSet.toList poss)
                                        (IntMap.keysSet degs)
              ndegs = IntMap.union (IntMap.fromSet (const i) nposs) degs

maxDepth = IntMap.foldr max 0 positionsBFS -- 33

retrievedPaths :: [[Position]]
retrievedPaths = map (map indexToPos) (go (maxDepth - 1) [[finalIndex]]) where
    go depth paths
        | depth == 0 = paths
        | otherwise = go (depth - 1) [nbi:path
            | path@(h:_) <- paths
            , nbi <- filter (\i -> positionsBFS IntMap.!? i == Just (depth - 1)) (neighboursIndexes h)
            ]

getPathVal :: [Position] -> Int
getPathVal = foldl updateChecksum 0 . foldDescend findMove

euler244 :: IO String
euler244 = return $ show $ sum $ map getPathVal retrievedPaths

File: rust/src/solutions/euler244.rs
use std::collections::{HashMap, VecDeque};

/*
-------------
| 0| 1| 2| 3|
-------------
| 4| 5| 6| 7|
-------------
| 8| 9|10|11|
-------------
|12|13|14|15|
-------------

nodes ~ positions             positions of blank
vertices ~ moves               ||
Number of vertices in graph is 16 * binom(15, 7) = 102960
Each node has at most 4 edges            ||
(directions to move for blank)           positions of red tiles
*/

#[derive(Debug)]
struct Position {
    blank: u32,
    reds: [u32; 7], // it is not required for position to be sorted
                    // blues: remaining ones
}

#[derive(Copy, Clone, Debug)]
enum Direction {
    Left,
    Right,
    Up,
    Down,
}

impl Direction {
    fn to_ascii_code(&self) -> u64 {
        match &self {
            Direction::Left => 76,
            Direction::Right => 82,
            Direction::Up => 85,
            Direction::Down => 68,
        }
    }
}

// Encoding of position as index:
// first 4 bits are used to represent position of blank,
// next 16 bits are used to represent positions of reds,

impl Position {
    fn to_index(&self) -> u32 {
        self.blank + (self.reds.iter().fold(0, |acc, &r| acc + (1 << r)) << 4)
    }

    fn neighbours(&self) -> Vec<(Direction, Position)> {
        let move_left = if self.blank & 3 == 3 {
            None
        } else {
            let blank = self.blank + 1;
            let mut reds = self.reds;
            for i in 0..7 {
                if reds[i] == blank {
                    reds[i] = blank - 1;
                }
            }
            Some((Direction::Left, Position { blank, reds }))
        };
        let move_right = if self.blank & 3 == 0 {
            None
        } else {
            let blank = self.blank - 1;
            let mut reds = self.reds;
            for i in 0..7 {
                if reds[i] == blank {
                    reds[i] = blank + 1;
                }
            }
            Some((Direction::Right, Position { blank, reds }))
        };
        let move_up = if self.blank >= 12 {
            None
        } else {
            let blank = self.blank + 4;
            let mut reds = self.reds;
            for i in 0..7 {
                if reds[i] == blank {
                    reds[i] = blank - 4;
                }
            }
            Some((Direction::Up, Position { blank, reds }))
        };
        let move_down = if self.blank < 4 {
            None
        } else {
            let blank = self.blank - 4;
            let mut reds = self.reds;
            for i in 0..7 {
                if reds[i] == blank {
                    reds[i] = blank + 4;
                }
            }
            Some((Direction::Down, Position { blank, reds }))
        };
        vec![move_left, move_right, move_up, move_down]
            .into_iter()
            .filter_map(|x| x)
            .collect()
    }

    fn init_position() -> Position {
        Position {
            blank: 0,
            reds: [1, 4, 5, 8, 9, 12, 13],
        }
    }

    fn final_position() -> Position {
        Position {
            blank: 0,
            reds: [2, 5, 7, 8, 10, 13, 15],
        }
    }
}

pub fn euler244() -> String {
    let mut queue = VecDeque::<Position>::new();
    queue.push_back(Position::init_position());
    let mut shortest_paths = HashMap::<u32, (usize, Vec<Vec<Direction>>)>::new();
    shortest_paths.insert(Position::init_position().to_index(), (0, vec![vec![]]));
    let final_index = Position::final_position().to_index();
    let mut final_state_visited = false;
    while !queue.is_empty() {
        let pos = queue.pop_front().unwrap();
        let pos_index = pos.to_index();
        let (shortest_path_len, paths) = shortest_paths[&pos_index].clone();
        if pos_index == final_index {
            final_state_visited = true;
        }
        for (direction, neighbour) in pos.neighbours() {
            let neighbour_index = neighbour.to_index();
            if !shortest_paths.contains_key(&neighbour_index) && !final_state_visited {
                queue.push_back(neighbour);
            }
            let entry = shortest_paths
                .entry(neighbour_index)
                .or_insert((shortest_path_len + 1, vec![]));
            if shortest_path_len + 1 == entry.0 {
                for path in &paths {
                    let mut new_path = path.clone();
                    new_path.push(direction);
                    entry.1.push(new_path);
                }
            }
        }
    }
    let (_, paths_to_final_position) = shortest_paths.get(&final_index).unwrap();
    paths_to_final_position
        .into_iter()
        .map(|path| {
            path.into_iter().fold(0, |checksum, direction| {
                (checksum * 243 + direction.to_ascii_code()) % 100_000_007
            })
        })
        .sum::<u64>()
        .to_string()
}
